featurebyte/service/feature_table_cache.py

```diff
@@ -1,111 +1,367 @@
+                )
+                for column in new_columns
+            ],
+        )
+        sql = sql_to_string(alter_expr, source_type=db_session.source_type)
+        await db_session.execute_query(sql)
 
-        return documents[0]
+        # merge temp table into cache table
+        merge_conditions = [
+            expressions.EQ(
+                this=get_qualified_column_identifier(
+                    InternalName.TABLE_ROW_INDEX, "faeture_table_cache"
+                ),
+                expression=get_qualified_column_identifier(
+                    InternalName.TABLE_ROW_INDEX, "partial_features"
+                ),
+            )
+        ]
+        existing_columns = [f"FEATURE_{definition_hash}" for definition_hash in cached_nodes.keys()]
+        update_expr = expressions.Update(
+            expressions=[
+                expressions.EQ(
+                    this=get_qualified_column_identifier(column, "faeture_table_cache"),
+                    expressions=get_qualified_column_identifier(column, "partial_features"),
+                )
+                for column in existing_columns
+            ]
+        )
+        insert_expr = expressions.Insert(
+            this=expressions.Tuple(
+                expressions=[quoted_identifier(column) for column in new_columns]
+            ),
+            expression=expressions.Tuple(
+                expressions=[
+                    get_qualified_column_identifier(column, "partial_features")
+                    for column in new_columns
+                ]
+            ),
+        )
+        merge_expr = expressions.Merge(
+            this=expressions.Table(
+                this=quoted_identifier(cache_metadata.table_name),
+                alias=expressions.TableAlias(this="feaeture_table_cache"),
+            ),
+            using=expressions.Table(
+                this=quoted_identifier(temp_final_cache_table_name),
+                alias=expressions.TableAlias(this="partial_features"),
+            ),
+            on=expressions.and_(*merge_conditions),
+            expressions=[
+                expressions.When(
+                    this=expressions.Column(this=expressions.Identifier(this="MATCHED")),
+                    then=update_expr,
+                ),
+                expressions.When(
+                    this=expressions.Not(
+                        this=expressions.Column(this=expressions.Identifier(this="MATCHED")),
+                    ),
+                    then=insert_expr,
+                ),
+            ],
+        )
+        sql = sql_to_string(merge_expr, source_type=db_session.source_type)
+        await db_session.execute_query(sql)
 
-    async def update_feature_table_cache(
+    async def create_or_update_feature_table_cache(
         self,
-        observation_table_id: PydanticObjectId,
-        feature_definitions: List[CachedFeatureDefinition],
+        feature_store: FeatureStoreModel,
+        observation_table: ObservationTableModel,
```

Comment:
Would be good to have an assertion that `observation_table.has_row_index` must be True when calling this function.